websocket/src/websocket.cpp

```diff
@@ -0,0 +1,536 @@
+    SetStatus(conn, RESULT_ERROR, __VA_ARGS__); \
+    CloseConnection(conn);
+
+
+static void SetState(WebsocketConnection* conn, State state)
+{
+    State prev_state = conn->m_State;
+    if (prev_state != state)
+    {
+        conn->m_State = state;
+        WS_DEBUG("%s -> %s", StateToString(prev_state), StateToString(conn->m_State));
+    }
+}
+
+
+Result SetStatus(WebsocketConnection* conn, Result status, const char* format, ...)
+{
+    if (conn->m_Status == RESULT_OK)
+    {
+        va_list lst;
+        va_start(lst, format);
+
+        conn->m_BufferSize = vsnprintf(conn->m_Buffer, conn->m_BufferCapacity, format, lst);
+        va_end(lst);
+        conn->m_Status = status;
+    }
+    return status;
+}
+
+// ***************************************************************************************************
+// LUA functions
+
+
+
+static WebsocketConnection* CreateConnection(const char* url)
+{
+    WebsocketConnection* conn = (WebsocketConnection*)malloc(sizeof(WebsocketConnection));
+    memset(conn, 0, sizeof(WebsocketConnection));
+    conn->m_BufferCapacity = g_Websocket.m_BufferSize;
+    conn->m_Buffer = (char*)malloc(conn->m_BufferCapacity);
+
+    dmURI::Parts uri;
+    dmURI::Parse(url, &conn->m_Url);
+
+    if (strcmp(conn->m_Url.m_Scheme, "https") == 0)
+        strcpy(conn->m_Url.m_Scheme, "wss");
+
+    conn->m_SSL = strcmp(conn->m_Url.m_Scheme, "wss") == 0 ? 1 : 0;
+    conn->m_State = STATE_CONNECTING;
+
+    return conn;
+}
+
+static void DestroyConnection(WebsocketConnection* conn)
+{
+#if defined(HAVE_WSLAY)
+    if (conn->m_Ctx)
+        WSL_Exit(conn->m_Ctx);
+#endif
+
+    if (conn->m_Callback)
+        dmScript::DestroyCallback(conn->m_Callback);
+
+    if (conn->m_Connection)
+        dmConnectionPool::Return(g_Websocket.m_Pool, conn->m_Connection);
+
+    free((void*)conn->m_Buffer);
+    free((void*)conn);
+}
+
+
+static void CloseConnection(WebsocketConnection* conn)
+{
+    State prev_state = conn->m_State;
+
+    // we want it to send this message in the polling
+    if (conn->m_State == STATE_CONNECTED) {
+#if defined(HAVE_WSLAY)
+        WSL_Close(conn->m_Ctx);
+#endif
+    }
+
+    SetState(conn, STATE_DISCONNECTED);
+}
+
+static int FindConnection(WebsocketConnection* conn)
+{
+    for (int i = 0; i < g_Websocket.m_Connections.Size(); ++i )
+    {
+        if (g_Websocket.m_Connections[i] == conn)
+            return i;
+    }
+    return -1;
+}
+
+/*#
+*
+*/
+static int LuaConnect(lua_State* L)
+{
+    DM_LUA_STACK_CHECK(L, 1);
+
+    if (!g_Websocket.m_Initialized)
+        return DM_LUA_ERROR("The web socket module isn't initialized");
+
+    const char* url = luaL_checkstring(L, 1);
+
+    // long playedTime = luaL_checktable_number(L, 2, "playedTime", -1);
+    // long progressValue = luaL_checktable_number(L, 2, "progressValue", -1);
+    // char *description = luaL_checktable_string(L, 2, "description", NULL);
+    // char *coverImage = luaL_checktable_string(L, 2, "coverImage", NULL);
```

Comment:
A bit silly to leave here (as documentaiton), but I suspect that we want to have configuration options soon enough.